### PR TITLE
[SC-902] Limit order v4 description

### PR DESCRIPTION
**Pull request by galekseev**
Docs and examples for Limit Order v4

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,69 +9,77 @@
...
-3) Callback, for to notify maker on order execution.
 
-### RFQ order
+The key features of the protocol are **extreme flexibility** and **high gas efficiency**, which is achieved with the following features
+
+**Basic features**
+
+- Select an asset receiver for an order.
+- Choose whether to allow or disallow partial and multiple fills.
+- Define conditions that must be met before execution can proceed (e.g. provide an example).
```
Comment (byshape) [inline README.md:34]: > e.g. provide an example

Looks like an example is missing

File: description.md [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+| UNWRAP_WETH | 247 | If set, the order requires unwrapping WETH |
+
+The rest of the settings are located in the lowest 200 bits of the number, from lowest to highest.
+
+| Option name | Size, bits | Description |
+| --- | --- | --- |
+| ALLOWED_SENDER | 80 | This option is used to make an order private and restrict filling to only one specified taker address. The option consists of the last 10 bytes of the allowed sender's address. A value of zero means that no restrictions will be applied. |
+| EXPIRATION | 40 | Expiration timestamp for the order. Order cannot be filled after the expiration deadline. Zero value means that there is no expiration time for the order. |
+| NONCE_OR_EPOCH | 40 | The nonce or epoch of the order. See <cancel order> for details. |
+| SERIES | 40 | The series of the order. See <cancel order> for details. |
```
Comment (zZoMROT) [inline description.md:105]:
```suggestion
| Option name | Size, bits | Position, bit numbers | Description |
| --- | --- | --- |
| ALLOWED_SENDER | 80 | 0-79 | This option is used to make an order private and restrict filling to only one specified taker address. The option consists of the last 10 bytes of the allowed sender's address. A value of zero means that no restrictions will be applied. |
| EXPIRATION | 40 | 80-119 | Expiration timestamp for the order. Order cannot be filled after the expiration deadline. Zero value means that there is no expiration time for the order. |
| NONCE_OR_EPOCH | 40 | 120-159 | The nonce or epoch of the order. See <cancel order> for details. |
| SERIES | 40 | 160-199 | The series of the order. See <cancel order> for details. |
```

File: description.md [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+Examples of tokens that have already implemented this method are based on the following standards:
+
+- ERC721 - `IERC721.transferFrom(from, to, tokenId)`
+- ERC1155 - `IERC1155.safeTransferFrom(from, to, tokenId, amount, data)`
+
+To implement your custom proxy:
+
+1. Implement a function that receives `from`, `to`, and `amount` as the first three parameters, along with additional parameters that will be passed with `makerAssetSuffix` and/or `takerAssetSuffix`.
+2. Find the function name whose selector equals `transferFrom(from,to,amount)`.
+3. Implement `transferFrom` for the token.
```
Comment (zZoMROT) [inline description.md:236]: It looks like this items should be a little more detailed. Where the one should search function with specific selector? Does `transferFrom` from item 3 is equal to the method from item 2?

File: description.md [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+
+Examples of tokens that have already implemented this method are based on the following standards:
+
+- ERC721 - `IERC721.transferFrom(from, to, tokenId)`
+- ERC1155 - `IERC1155.safeTransferFrom(from, to, tokenId, amount, data)`
+
+To implement your custom proxy:
+
+1. Implement a function that receives `from`, `to`, and `amount` as the first three parameters, along with additional parameters that will be passed with `makerAssetSuffix` and/or `takerAssetSuffix`.
+2. Find the function name whose selector equals `transferFrom(from,to,amount)`.
```
Comment (zZoMROT) [inline description.md:235]:
```suggestion
2. Compute and select a function name whose selector matches `transferFrom(from,to,amount)`.
```

File: description.md (hunk 3 above)
Comment (zZoMROT) [inline description.md:236]:
```suggestion
3. Implement your own `transferFrom` method in the function from item 2 for the token
```

File: description.md [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
+1. Include the proxy address in the `makerAsset` or `takerAddress` field.
+2. Have the `HAS_EXTENSION` flag set.
+3. Include an extension with either `makerAssetSuffix` or `takerAssetSuffix` set, with the packed `tokenId` and `token` address. This information will be passed to the proxy when the order is filled.
+
+> Note: This extension is incompatible with the `USE_PERMIT2_FLAG` setting.
+> 
+
+**Example**
+
+Here is the code that creates an order:
```
Comment (zZoMROT) [inline description.md:272]:
```suggestion
Here is the sample code that creates an order:
```

File: description.md [hunk 6]
```diff
@@ -0,0 +1,937 @@
...
+
+```javascript
```
Comment (zZoMROT) [inline description.md:274]: Maybe we should rename `addr1` in example and explain where users can get `buildOrder` method

File: description.md [hunk 7]
```diff
@@ -0,0 +1,937 @@
...
+The code creates an order that uses `RangeAmountCalculator` to calculate the making and taking amounts.
+
+```javascript
+// Order: 10 weth -> 35000 dai with price range: 3000 -> 4000
+const makingAmount = ether('10');
+const takingAmount = ether('35000');
+const startPrice = ether('3000');
+const endPrice = ether('4000');
+
+const makingAmountGetter = rangeAmountCalculator.address + trim0x(cutLastArg(cutLastArg(
```
Comment (zZoMROT) [inline description.md:383]: Should we explain that `rangeAmountCalculator` is object  of deployed contract `RangeAmountCalculator` or is it redundant information?

File: description.md [hunk 8]
```diff
@@ -0,0 +1,937 @@
...
+    rangeAmountCalculator.interface.encodeFunctionData('getRangeMakerAmount', [startPrice, endPrice, makingAmount, 0, 0], 64),
+)));
+
+const takingAmountGetter = rangeAmountCalculator.address + trim0x(cutLastArg(cutLastArg(
+    rangeAmountCalculator.interface.encodeFunctionData('getRangeTakerAmount', [startPrice, endPrice, makingAmount, 0, 0], 64),
```
Comment (zZoMROT) [inline description.md:388]: Do we really need `cutLastArg`x2? Is it useful to know why we cut something?

File: description.md [hunk 9]
```diff
@@ -0,0 +1,937 @@
...
+
+The code creates an order that uses `RangeAmountCalculator` to calculate the making and taking amounts.
+
+```javascript
+// Order: 10 weth -> 35000 dai with price range: 3000 -> 4000
+const makingAmount = ether('10');
+const takingAmount = ether('35000');
+const startPrice = ether('3000');
+const endPrice = ether('4000');
+
```
Comment (zZoMROT) [inline description.md:382]:
```suggestion

const rangeAmountCalculator = await ethers.getContractFactory('RangeAmountCalculator');

```

File: test/examples/LimitOrderProtocol-example.js [hunk 10]
```diff
@@ -0,0 +1,284 @@
+const hre = require('hardhat');
+const { ethers } = hre;
+const { expect, time, constants, trim0x } = require('@1inch/solidity-utils');
+const { fillWithMakingAmount, buildMakerTraits, buildOrder, signOrder, compactSignature, ABIOrder } = require('../helpers/orderUtils');
+const { loadFixture } = require('@nomicfoundation/hardhat-network-helpers');
+const { joinStaticCalls, ether, cutLastArg } = require('../helpers/utils');
+
+describe.skip('LimitOrderProtocol usage example', function () {
```
Comment (zZoMROT) [inline test/examples/LimitOrderProtocol-example.js:8]: Is this by design (`.skip`)?

File: description.md [hunk 11]
```diff
@@ -0,0 +1,937 @@
...
+| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| [28..31] | [24..27] | [20..23] | [16..19] | [12..15] | [8..11] | [4..7] | [0..3] | [32..151] | [152..183] | [184..216] |
+| 152 | 152 | 152 | 120 | 0 | 0 | 0 | 0 | calldata (120 bytes) | calldata (32 bytes) | calldata (32 bytes) |
+
+The following is the final calldata which includes the offsets.
+
+```bash
+#  28-31    24-27    20-23    16-19    12-15     8-11      4-7      0-3
+00000098 00000098 00000098 00000078 00000000 00000000 00000000 00000000
+#         predicate end => ^^^^^^^^ ^^^^^^^^ <= predicate start
```
Comment (zZoMROT) [inline description.md:212]: Above in line 191 it is said that the offset of predicate is in 16..19 bits

File: description.md [hunk 12]
```diff
@@ -0,0 +1,937 @@
...
+#   permit end => ^^^^^^^^ ^^^^^^^^ <= permit start
```
Comment (zZoMROT) [inline description.md:213]: The offset data for the `predicate` and `permit` seem to overlap

File: description.md [hunk 13]
```diff
@@ -0,0 +1,937 @@
...
+        bytes calldata extension
+    ) public returns(uint256 makingAmount, uint256 takingAmount, bytes32 orderHash)
+```
+
+| Argument | Type | Description |
+| --- | --- | --- |
+| order | `Order` (calldata) | The order structure to be filled.  |
+| r | `bytes32` | The r-component of the maker’s signature to check that the order hash is signed by the maker. |
+| vs | `bytes32` | The vs-component of the maker’s signature to check that the order hash is signed by the maker. |
+| signature | `bytes calldata` | The signature used to verify the order. It is used for contract-signed orders only. |
```
Comment (zZoMROT) [inline description.md:878]: It could be useful to mention ERC-1271 and `isValidSignature` method

File: description.md [hunk 14]
```diff
@@ -0,0 +1,937 @@
...
+| amount | `uint256` | The amount to fill the order, which can be treated as the maker or taker amount, depending on fill settings. If the amount is greater than the remaining amount to fill, the fill will be executed only for the remaining amount. When partial fills are not allowed, the fill will be reverted if the amount does not equal the order making amount. The fill will also be reverted if the making and taking amounts are equal to zero. |
```
Comment (zZoMROT) [inline description.md:879]: Small update about `making and taking amounts are equal to zero`
```
unchecked { if (makingAmount * takingAmount == 0) revert SwapWithZeroAmount(); }
```

File: description.md [hunk 15]
```diff
@@ -0,0 +1,937 @@
...
+| takerTraits | `TakerTraits` (uint256) | The taker’s setting for the order fill. See Fill settings for details. |
+| target | `address` | The recipient address for maker assets transfer.. |
```
Comment (zZoMROT) [inline description.md:881]:
```suggestion
| target | `address` | The recipient address for maker assets transfer. |
```

File: description.md [hunk 16]
```diff
@@ -0,0 +1,937 @@
...
+| interaction | `bytes calldata` | The taker interaction to execute during the fill. See interactions for details. |
```
Comment (zZoMROT) [inline description.md:882]:
```suggestion
| interaction | `bytes calldata` | The taker interaction to execute during the fill. See [Interactions](#interactions) section for details. |
```

File: description.md [hunk 17]
```diff
@@ -0,0 +1,937 @@
...
+| extension | `bytes calldata` | The order’s extension calldata. The extension’s keccak256 hash has to be equal to the 160-lower bit of order’s salt. |
+| permit | `bytes calldata` | The taker’s permit for the taker assets transfer. |
+
+The return values are:
+
+| Return value | Type | Description |
+| --- | --- | --- |
+| makingAmount | `uint256` | The actual amount the maker received  |
+| takingAmount | `uint256` | The actual amount the taker recieved |
```
Comment (zZoMROT) [inline description.md:891]:
```suggestion
| takingAmount | `uint256` | The actual amount the taker received |
```

File: description.md [hunk 18]
```diff
@@ -0,0 +1,937 @@
...
+### Manually (direct ways)
+
+Manual methods require sending a cancel transaction, which requires spending gas.
+
+- **Cancel by hash:** The order can be cancelled by directly calling the `cancelOrder` function and passing the `orderHash` and `makerTraits` of the order.
+
+    > **Note**: Orders are cancelled using different invalidators depending on the maker traits flags `ALLOW_MULTIPLE_FILLS` and `NO_PARTIAL_FILL`. Passing wrong traits may result in the call having no effect, and the order will not be cancelled.
+    > 
+
+    The fill attempts will be reverted with a `BitInvalidatedOrder` error if an order doesn't allow either partial or multiple fills, or with an `InvalidatedOrder` error otherwise.
```
Comment (zZoMROT) [inline description.md:924]: This is not very clear what this is referring to

File: test/examples/LimitOrderProtocol-example.js (hunk 10 above)
Comment (galekseev) [reply test/examples/LimitOrderProtocol-example.js:8]: Yes, this is an example of usage. So these tests  are not intended to run on each pass.

File: description.md (hunk 6 above)
Comment (galekseev) [reply description.md:274]: These examples are located in tests/examples/LimitOrderProtocol-example.js. If a reader needs it, he can find the implementation in code.

File: description.md (hunk 7 above)
Comment (galekseev) [reply description.md:383]: Here the focus is on the calldata produced by the code. A user can implement any code that produces the same calldata. The point is that calldata should be the same as shown below

File: description.md (hunk 8 above)
Comment (galekseev) [reply description.md:388]: Again, the focus is on the calldata produced by the code. A user can implement any code that produces the same calldata. The point is that calldata should be the same as shown below

File: README.md (hunk 1 above)
Comment (galekseev) [reply README.md]: fixed

File: description.md (hunk 3 above)
Comment (galekseev) [reply description.md]: fixed

File: description.md (hunk 11 above)
Comment (galekseev) [reply description.md]: fixed

File: description.md (hunk 13 above)
Comment (galekseev) [reply description.md]: fixed

File: description.md (hunk 14 above)
Comment (galekseev) [reply description.md]: fixed

**Review by byshape — CHANGES_REQUESTED** [review]
It would be nice to fix some typos before merging.

File: description.md (hunk 18 above)
Comment (galekseev) [reply description.md]: fixed

File: description.md [hunk 19]
```diff
@@ -0,0 +1,937 @@
...
+> **Note:** The order becomes invalidated after it is filled. Therefore, the salt should be different for orders that have all parameters equal. This ensures that the fill of one order does not invalidate all orders.
+> 
+
+## Order settings
+
+The `makerTraits` property contains order settings as bit flags and compacted numbers in a `uint256` format. The bit flags are listed from highest to lowest bit.
+
+| Option name | Bit position | Description |
+| --- | --- | --- |
+| NO_PARTIAL_FILLS | 255 | If set, the order does not allow partial fills.<br/>Partial fill only occurs when part of the required maker’s asset amount is swapped. This could be useful for large orders that can be hardly filled by only one taker. |
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:88]: yes, exactly

File: description.md [hunk 20]
```diff
@@ -0,0 +1,939 @@
...
+| USE_PERMIT2 | 248 | If set, the order uses [Uniswap Permit2](https://github.com/Uniswap/permit2)  |
+| UNWRAP_WETH | 247 | If set, the order requires unwrapping WETH |
+
+The rest of the settings are located in the lowest 200 bits of the number, from lowest to highest.
+
+| Option name | Size, bits | Location, bits | Description |
+| --- | --- | --- | --- |
+| ALLOWED_SENDER | 80 | [0..79] | This option is used to make an order private and restrict filling to only one specified taker address. The option consists of the last 10 bytes of the allowed sender's address. A value of zero means that no restrictions will be applied. |
+| EXPIRATION | 40 | [80..119] | Expiration timestamp for the order. Order cannot be filled after the expiration deadline. Zero value means that there is no expiration time for the order. |
+| NONCE_OR_EPOCH | 40 | [120..159] | The nonce or epoch of the order. See <cancel order> for details. |
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:104]: fixed

File: description.md [hunk 21]
```diff
@@ -0,0 +1,939 @@
...
+const LimitOrderProtocol = await ethers.getContractFactory('LimitOrderProtocol');
+const swap = await LimitOrderProtocol.deploy(weth.address);
+await swap.deployed();
+// Build predicate
+// 1. Create calldata to call oracle to get latest DAI price
+const priceCall = swap.interface.encodeFunctionData('arbitraryStaticCall', [
+            daiOracle.address,
+            daiOracle.interface.encodeFunctionData('latestAnswer'),
+        ]);
+// 2. Create calldata for condition that price should be less then 1000
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:531]: fixed

File: description.md [hunk 22]
```diff
@@ -0,0 +1,939 @@
...
+const comparelt = swap.interface.encodeFunctionData('lt', [ether('1000'), priceCall])
+// 3. Create calldata for condition that price should be more then 2000
+const comparegt = swap.interface.encodeFunctionData('gt', [ether('2000'), priceCall])
+// 4. Add `or` condition (joinStaticCalls - concatenates calldata)
+const { offsets, data } = joinStaticCalls([comparelt, comparegt]);
+**predicate** = swap.interface.encodeFunctionData('or', [offsets, data]);
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:537]: fixed

File: description.md [hunk 23]
```diff
@@ -0,0 +1,939 @@
...
+```
+
+| Argument | Type | Description |
+| --- | --- | --- |
+| order | `Order` (calldata) | The order structure to be filled.  |
+| r | `bytes32` | The r-component of the maker’s signature to check that the order hash is signed by the maker. |
+| vs | `bytes32` | The vs-component of the maker’s signature to check that the order hash is signed by the maker. |
+| signature | `bytes calldata` | The signature used to verify the order. It is used for contract-signed orders only. |
+| amount | `uint256` | The amount to fill the order, which can be treated as the maker or taker amount, depending on fill settings. If the amount is greater than the remaining amount to fill, the fill will be executed only for the remaining amount. When partial fills are not allowed, the fill will be reverted if the amount does not equal the order making amount. The fill will also be reverted if the making and taking amounts are equal to zero. |
+| takerTraits | `TakerTraits` (uint256) | The taker’s setting for the order fill. See Fill settings for details. |
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:884]: fixed

File: description.md [hunk 24]
```diff
@@ -0,0 +1,939 @@
...
+| takingAmount | `uint256` | The actual amount the taker received |
+| orderHash | `bytes32` | The hash of the order |
+
+## Fill settings
+
+The `takerTraits` argument provides a number of options for the taker to choose from during each fill. These options are stored as bit flags and numbers compacted in a `uint256` number. The bit flags in `takerTraits` are arranged in descending order of significance, with the highest bit first.
+
+| Option name | Bit position | Description |
+| --- | --- | --- |
+| MAKER_AMOUNT_FLAG | 255 bit | If set, the protocol implies that the passed amount is the making amount, and the taking amount will be calculated based on the making amount. Otherwise, the passed amount is the taking amount, and the making amount is calculated based on the taking amount. The amount is calculated with AmountCalculator if getters are not set, or with getters provided with an extension by the maker. |
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:904]: yes, exactly

File: description.md [hunk 25]
```diff
@@ -0,0 +1,941 @@
...
+- **Cancel by hash:** The order can be cancelled by directly calling the `cancelOrder` function and passing the `orderHash` and `makerTraits` of the order.
+
+    > **Note**: Orders are cancelled using different invalidators depending on the maker traits flags `ALLOW_MULTIPLE_FILLS` and `NO_PARTIAL_FILL`. Passing wrong traits may result in the call having no effect, and the order will not be cancelled.
+    > 
+
+    The fill attempts will be reverted with a `BitInvalidatedOrder` error if an order doesn't allow either partial or multiple fills, or with an `InvalidatedOrder` error otherwise.
+
+- **Cancel by nonce:** the order is cancelled by changing the order nonce. This method can be used for mass order cancellation. Each order can have a series and nonces specified. They are defined as:
+    - **series** - specifies the application that issued the order
+    - **nonce** - ****specifies the order’s generation
```
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:933]: fixed

File: description.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (byshape) [reply description.md:88]: Then I think it's worth mentioning this fact (counting from 0), what do you think?

File: description.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (byshape) [reply description.md:104]: Could you please give me a link to this fix? In the PR files it's still the same for me
image

File: description.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (byshape) [reply description.md]: Please, fix this as well `Create calldata for condition that price should be more then 2000`

File: description.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:104]: Sorry, fixed it in line 93 and missed it here.
Now it should be fixed.

File: description.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md:88]: fixed

File: description.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (galekseev) [reply description.md]: fixed

**Review by byshape — APPROVED** [review]

**Pull request merged**